src/prpy/tsr/tsr.py

```diff
@@ -52,36 +61,132 @@ def __init__(self, T0_w=None, Tw_e=None, Bw=None,
-            raise ValueError('vals must be of length %d or 6!' % Bwdims)
-        # print 'Bwvals[5]:', Bwvals[5]
-        xyzypr = [Bwvals[0], Bwvals[1], Bwvals[2],
-                  Bwvals[5], Bwvals[4], Bwvals[3]]
-        Tw = kin.pose_to_H(kin.pose_from_xyzypr(xyzypr))
+    def to_transform(self, xyzrpy):
+        """
+        Converts a [x y z roll pitch yaw] into an
+        end-effector transform.
 
+        @param  xyzrpy [x y z roll pitch yaw]
+        @return trans 4x4 transform
+        """
+        if len(xyzrpy) != 6:
+            raise ValueError('vals must be of length 6')
+        if not self.is_valid(xyzrpy):
+            raise ValueError('Invalid xyzrpy')
+
+        xyzypr = [xyzrpy[0], xyzrpy[1], xyzrpy[2],
+                  xyzrpy[5], xyzrpy[4], xyzrpy[3]]
+        Tw = kin.pose_to_H(kin.pose_from_xyzypr(xyzypr))
         trans = numpy.dot(numpy.dot(self.T0_w, Tw), self.Tw_e)
         return trans
 
+    def to_xyzrpy(self, trans):
+        """
+        Converts an end-effector transform to xyzrpy values
+        @param  trans  4x4 transform
+        @return xyzrpy 6x1 vector of Bw values
+        """
+        Tw = numpy.dot(numpy.dot(numpy.linalg.inv(self.T0_w), trans),
+                       numpy.linalg.inv(self.Tw_e))
+        pose = kin.pose_from_H(Tw)
+        ypr = kin.quat_to_ypr(pose[3:7])
+        xyzrpy = [pose[0], pose[1], pose[2],
+                  ypr[2], ypr[1], ypr[0]]
+        return xyzrpy
+
+    def is_valid(self, xyzrpy, ignoreNAN=False):
+        """
+        Checks if a xyzrpy is a valid sample from the TSR.
+        Two main issues: dealing with roundoff issues for zero bounds and
+        Wraparound for rpy.
+        @param xyzrpy 6x1 vector of Bw values
+        @param ignoreNAN (optional, defaults to False) ignore NaN xyzrpy
+        @return True if valid and False if not
+        """
+        Bw_xyz = self.Bw[0:3, :]
+
+        xyzcheck = [(((x + EPSILON) >= Bw_xyz[i, 0]) and 
+                     ((x - EPSILON) <= Bw_xyz[i, 1]))
+                    or (Bw_xyz[i, 1] - Bw_xyz[i, 0] < EPSILON)
+                    for i, x in enumerate(xyzrpy[0:3])]
+
+        # Unwrap all rotations to [-pi, pi]
+        Bw_rpy = (self.Bw[3:6, :] + pi) % (2*pi) - pi
+        rpy = numpy.add(xyzrpy[3:6], pi) % (2*pi) - pi
+        rpycheck = [(((x + EPSILON) >= Bw_rpy[i, 0]) and 
+                     ((x - EPSILON) <= Bw_rpy[i, 1]))
+                    or (Bw_rpy[i, 1] - Bw_rpy[i, 0] < EPSILON)
+                    for i, x in enumerate(rpy)]
+
+        check = numpy.hstack((xyzcheck, rpycheck))
+
+        # Strip out Nans if needed
+        if ignoreNAN:
+            check = check[~numpy.isnan(xyzrpy)]
+
+        return all(check)
+
+    def contains(self, trans):
+        """
+        Checks if the TSR contains the transform
+        @param  trans 4x4 transform
+        @return       True if inside and False if not
+        """
+        xyzrpy = self.to_xyzrpy(trans)
+        return self.is_valid(xyzrpy)
```

Comment:
Instead of a `bool`, it might be nice to return the mask. This could help us write very informative error messages about why a constraint is being violated.